File: javascript/matrix/1260. Shift 2D Grid.js
/**
 * @param {number[][]} grid
 * @param {number} k
 * @return {number[][]}
 */
var shiftGrid = function(grid, k) {
    const M = grid.length, N = grid[0].length;

    const posToIdx = (row, col) => {
        return row * N + col;
    }
    const idxToPos = (idx) => {
        return [Math.floor(idx / N), idx % N];  // can only return 1 value
    }

    let res = Array(M).fill().map(() => Array(N));
    for (let r=0; r<M; r++) {
        for (let c=0; c<N; c++) {
            let newIdx = (posToIdx(r, c) + k) % (M * N);
            let [newR, newC] = idxToPos(newIdx);
            res[newR][newC] = grid[r][c];
        }
    }
    return res;
};

File: python/math/1260. Shift 2D Grid.py
'''
2D -> 1D -> 2D
Time: O(m * n)
Space: O(m * n)
'''
class Solution:
    def shiftGrid(self, grid: List[List[int]], k: int) -> List[List[int]]:
        M, N = len(grid), len(grid[0])

        def TwoD2OneD(row, col):
            return row * N + col
        def OneD2TwoD(idx):
            return idx // N, idx % N

        res = [[0] * N for i in range(M)]  # new m*n grid
        
        for r in range(M):
            for c in range(N):
                newIdx1D = (TwoD2OneD(r, c) + k) % (M * N)  # %(M*N): edge case
                newR, newC = OneD2TwoD(newIdx1D)
                res[newR][newC] = grid[r][c]

        return res